### PR TITLE
Option to listen to postRemove event

### PR DESCRIPTION
**Pull request by codedge**
This PR adds the option to listen to Doctrines `postRemove`. This is a result of the discussion in #62, a first step.

You can know specify in `meili_search.yml`:

```
meili_search:
    # One, or multiple of postPersist, postUpdate, preRemove, postRemove
    doctrineSubscribedEvents: ['postPersist', 'postUpdate', 'preRemove', 'postRemove'] 
```

@curquiza When merging this, we need to update the [wiki - Indexing automaticlly via ...](https://github.com/meilisearch/meilisearch-symfony/wiki/index-data-into-meilisearch#indexing-automatically-via-doctrine-events)

### CODE REVIEW

**Comment by norkunas** [comment]
Now I remember: the things is that doctrine before dispatching `postRemove` event unsets the primary key so there is an error:
```
In SearchableEntity.php line 85:
                                                             
  [Symfony\Component\Config\Definition\Exception\Exception]  
  Entity has no primary key                                  
                                                             

Exception trace:
  at /var/www/html/vendor/meilisearch/search-bundle/src/SearchableEntity.php:85
 MeiliSearch\Bundle\SearchableEntity->setId() at /var/www/html/vendor/meilisearch/search-bundle/src/SearchableEntity.php:52
 MeiliSearch\Bundle\SearchableEntity->__construct() at /var/www/html/vendor/meilisearch/search-bundle/src/Services/MeiliSearchService.php:337
 MeiliSearch\Bundle\Services\MeiliSearchService->makeSearchServiceResponseFrom() at /var/www/html/vendor/meilisearch/search-bundle/src/Services/MeiliSearchService.php:131
 MeiliSearch\Bundle\Services\MeiliSearchService->remove() at /var/www/html/vendor/meilisearch/search-bundle/src/EventListener/MeiliSearchIndexerSubscriber.php:47
 MeiliSearch\Bundle\EventListener\MeiliSearchIndexerSubscriber->postRemove() at /var/www/html/vendor/symfony/doctrine-bridge/ContainerAwareEventManager.php:68
 Symfony\Bridge\Doctrine\ContainerAwareEventManager->dispatchEvent() at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/Event/ListenersInvoker.php:115
 Doctrine\ORM\Event\ListenersInvoker->invoke() at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:1251
 Doctrine\ORM\UnitOfWork->executeDeletions() at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:446
 Doctrine\ORM\UnitOfWork->commit() at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php:378
 Doctrine\ORM\EntityManager->flush() at /var/www/html/var/cache/dev/Container52kWrqR/EntityManager_9a5be93.php:129
 Container52kWrqR\EntityManager_9a5be93->flush() at /var/www/html/src/Command/TestasCommand.php:70
 App\Command\TestasCommand->execute() at /var/www/html/vendor/symfony/console/Command/Command.php:299
 Symfony\Component\Console\Command\Command->run() at /var/www/html/vendor/symfony/console/Application.php:996
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/vendor/symfony/framework-bundle/Console/Application.php:96
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /var/www/html/vendor/symfony/console/Application.php:295
 Symfony\Component\Console\Application->doRun() at /var/www/html/vendor/symfony/framework-bundle/Console/Application.php:82
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /var/www/html/vendor/symfony/console/Application.php:167
 Symfony\Component\Console\Application->run() at /var/www/html/vendor/symfony/runtime/Runner/Symfony/ConsoleApplicationRunner.php:56
 Symfony\Component\Runtime\Runner\Symfony\ConsoleApplicationRunner->run() at /var/www/html/vendor/autoload_runtime.php:35
 require_once() at /var/www/html/bin/console:15
```

**Comment by norkunas** [comment]
Also I've forgot about rollbacked transactions :) but that probably should follow in another PR.

**Comment by codedge** [comment]
There are currently no tests for listening and then acting on the events coming from Doctrine. I'll add some to catch these kind of errors.

**Comment by curquiza** [comment]
Thanks for your PR @codedge, request me as reviewer when you think this is finished! 🙂

**Comment by codedge** [comment]
After some digging this is the result:

## `postRemove` has no entity id

The `postRemove` event is obviously not the best choice when you need the entity id for further processing. 
There is a big discussion about this here https://github.com/doctrine/orm/issues/2326 with no outcome. It is just by design.

Yes, we can tackle this with [some hacks](https://github.com/doctrine/orm/issues/2326#issuecomment-754732218), f. ex. saving the entity in some new field in the `preRemove` event so that we still have it in the `postRemove` event but that is... hacky. I am not a big fan of that.

The hack would mean that we always would need to listen to the `preRemove` event when the `postRemove` event is configured.

So using the `postRemove` event for removing a document from MS is likely not going to be successful. What dou you think?

## Transactions

If you @norkunas can outline what your concerns are regarding? 😄 

I would think if your transaction fails, meaning the removal of the entity in your db is not successful, but the `preRemove` event already fired and removed the document from MS - then why not adding it again?

It seems there are also new [Transaction events](https://github.com/doctrine/dbal/pull/4622) added in Doctrine so you could catch those errors and call MS again to re-index the document.

Let me know, where I am wrong here 😉

**Comment by norkunas** [comment]
My primary concern was to have at least integration with messenger by adding but the problem is that `SearchService` does not have a method to remove document by identifier which I could handle by myself when needed with preRemove setting the ID and postRemove dispatching the message :)

**Comment by codedge** [comment]
I see :D Well, then  we all learned a bit about Doctrine :D :D 

@curquiza 
Then I would still push some test for this one here and remove the `postRemove` config for now. Then we should move on with the messenger.

**Comment by codedge** [comment]
Changes in this PR are only pointing towards testing, code style (strict typings, format).

## PHPstan

I added https://github.com/phpstan/phpstan for checking the coding regarding type juggling.
There is a composer script `composer phpstan` that start and runs phpstan.

The current level is `5`, `8` are available and the more the better. For now, I think this is quite a good result.

@curquiza It would be nice if this could be a check (with bors) for all PRs and after merge.

## Separating tests

I felt like some tests are more like unit testing, most are integration tests. That's why I split them into a `Unit` and an `Integration` directory and created two test suites. That's nothing anyone will notice but it seems it gives some more structure to the testing files.

Further the unit tests can rely on the default `KernelTestCase` that is shipped with Symfony. I feel like sticking to the (Symfony) defaults is like a good thing.

## Code style

I added the new PHP CS Fixer rule [`PHP80MigrationRisky`](https://cs.symfony.com/doc/ruleSets/PHP80MigrationRisky.html) which works a bit more towards stricter code/types by adding return types to methods.

## Doctrine events tests & coverage

There are now doctrine event subscriber tests. The overall test coverage is about 85%. 
There is additionally a composer command `composer test:unit:coverage` which will create the code coverage (if you have either xdebug or pcov installed).

## Obosolete files

I found two files that are obsolete and used nowhere.

File: src/Engine.php [hunk 1]
```diff
@@ -80,7 +80,7 @@ public function remove($searchableEntities): array
         /** @var SearchableEntity $entity */
         foreach ($searchableEntities as $entity) {
             $searchableArray = $entity->getSearchableArray();
-            if (null === $searchableArray || 0 === \count($searchableArray)) {
+            if (null == $searchableArray || 0 == \count($searchableArray)) {
```
Comment (norkunas) [inline src/Engine.php:83]: Why this change? :)

Comment (codedge) [reply src/Engine.php:83]: This change came from PHPstan stating that this comparison is always `null`. I prepared an example in the PHPstan playground: https://phpstan.org/r/07c0aa63-1a31-4a24-b6af-c14dffa92dc5

**Review by curquiza — APPROVED** [review]
Thank you a lot everyone!!
And sorry for the delay!

bors merge

🚀

**Comment by codedge** [comment]
bors has its bad days 😃 

@curquiza Can you create a release out of the current master branch after this is merged? My last changes regarding the installation, #94, is thing I need for the Symfony recipe. The Symfony recipe checks fails, because of the error fixed in #94 .

**Comment by curquiza** [comment]
I merge without bors then, because I'm on holiday in few minutes during a month and I don't want to wait 😇
